### PR TITLE
feat: support metadata 2.1-2.3

### PR DESCRIPTION
**Pull request by henryiii**
Working on supporting metadata 2.3 and allowing a backend to choose what version they want.

In draft until I add tests.

### CODE REVIEW

**Review by frostming — APPROVED** [review]

File: pyproject_metadata/__init__.py [hunk 1]
```diff
@@ -229,6 +240,10 @@ def from_pyproject(
             msg = 'The description must be a single line'
             raise ConfigurationError(msg)
 
+        if metadata_version and metadata_version not in KNOWN_METADATA_VERSIONS:
```
Comment (dnicolodi) [inline pyproject_metadata/__init__.py:243]: The first condition in the `if` is redundant

Comment (frostming) [reply pyproject_metadata/__init__.py:243]: It's not, this is an error branch and `None` is a legal value.

Comment (dnicolodi) [reply pyproject_metadata/__init__.py:243]: Oh, you're right. I misread the code.

**Comment by henryiii** [comment]
I think I slightly misunderstood Metadata 2.3. I believe all extras before 2.3 are (now) supposed to be normalized, and 2.3 restricts the valid set to normalized extras only. Currently, we are producing:

```
Provides-Extra: under_score
Requires-Dist: some_package; extra == "under-score"
```

Which is incorrect, these don't match. It's caused by `packaging.markers.Marker` always normalizing extras. I think we should too.

I'm not sure of my reading of the PEP and guidelines, but happy to be corrected if wrong!

File: pyproject_metadata/__init__.py (hunk 1 above)
Comment (henryiii) [reply pyproject_metadata/__init__.py:243]: I'm trying to keep the original behavior, though I'm not that fond of it - setting a value as dynamic (PEP 621) doesn't mean it's dynamic SDist -> wheel. Lots of examples, like readme generation, are still static.

**Comment by henryiii** [comment]
@frostming I re-requested a review just to make sure you were okay with the extra normalization always occurring.

**Comment by henryiii** [comment]
pre-commit.ci autofix

**Review by frostming — APPROVED** [review]

**Pull request merged**